### PR TITLE
New Minor Releases (VM Dynamic GasCosts, EIP-2681, Sepolia, various other)

### LINKED ISSUE #1635: Monorepo: Next Minor Release Round
The next minor release round is at the doorstep, various libraries are up for releases with substantial feature additions, release roughly planned for within the next 1-3 weeks (base date: Saturday, January 15 2022). Depends a bit if there are last things to discuss or take in.

This is somewhat of a tracking issue for myself and also for making the process transparent and let people chime in if they have a request.

Currently I would plan for the following releases (just citing the most significant changes here for some overview, will properly link to the PRs in an upcoming release PR):

- RLP
  - Not sure: should there be a `v2.2.7` -> `v2.2.8` release? (mainly with https://github.com/ethereumjs/rlp/pull/109 I guess) 

- Trie `v4.2.2` -> `v4.2.3`
  - Deduped RLP dependency

- Util `v7.1.3` -> `v7.1.4`
  - RLP dedupe
  - validateNoLeadingZeroes()
  - new constants
  - negative BN checks

- Common `v2.6.0` -> `v2.6.1`
  - Sepolia
  - EIP-3855 (Push0)

- Tx `v3.4.0` -> `v3.5.0`
  - 2^64-1 value checks
  - EIP-2681
  - Sepolia
  - Tx integer leading zero checks
  - Throw on negative BN

- Block, Blockchain
  - Cautiously I would say no releases needed

- devp2p `v4.2.0` -> `v4.2.1`
  - Dedupe RLP dependency

- VM `v5.6.0` -> `v5.7.0`
  - Dynamic Gas Cost Refactoring
  - StateManager refactoring (BaseStateManager)
  - Proof State Manager (discuss `ProofStateManager` interface exposure!!)
  - EIP-2681 (Limit account nonce to 2^64-1)
  - EIP-3607 (Reject transactions from senders with deployed code)
  - EIP-3855 (Push0)
  - Sepolia
  - Tx out-of-gas consensus issue
  - Jump Dest Analysis Refactoring
  - Throw on negative BN
  
- Client `v0.2.0` -> `v0.3.0`
  - Client releases now always necessary (at least on minor), otherwise: side effects!
    - -> But: also makes sense in general! 🙂 
  - This (a client release) would need a general consistency check (maybe we should standardize a bit)!
  - ReceiptsManager, save receipts and logs (to re-confirm: this doesn't need a DB delete + fresh sync, right?)
  - Sepolia
  - Allow past block numbers in RPC queries
  - Fixed mainnet consensus bug (see VM)
  - Libp2p fixes
  - Error logging fixes/improvements
  
  Ok, so far for a high-level overview.
  
  Phew, this impresses me each and every time again how extensive this gets when writing this together. 😀
  
  Let me know if you agree with the extend of releases, the version numbers chosen (eventually some of this can rather be discussed in chat to not bloat here too much), also if you think I missed some major features to mention here (I didn't list all the smaller things yet) since then there might be a chance that I've really overlooked some older PR which took some time to finish.

### PR DESCRIPTION
**Pull request by holgerd77**
Fixes #1635 

Still WIP but libraries tackled already open for review.

Will likely remain open (and updated) for a couple of days (or maximally 1-2 weeks) for some final discussions and eventual inclusions (big thing: ESM support still addressable for release or not?).

### CODE REVIEW

**Comment by holgerd77** [comment]
Ah, sorry for all the "Build failed" notifications here, I will never learn this new "build package-lock.json file first" procedure. 😬

**Comment by holgerd77** [comment]
Ok, this is now ready for a first review, all libraries included! 😄

There will be some last work added in a last iteration, so if a PR of the last 2-3 days is still missing no need to mention.

Also, we might still add some selected additional work if decided.

**Comment by holgerd77** [comment]
(please ignore the outdated branch status, atm no need to keep completely updated yet)

**Comment by holgerd77** [comment]
Note to self: correct (remove) "Will error if there are uncommitted checkpoints on the instance." code doc comment on `StateManager.getStateRoot()`

File: packages/client/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,37 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+
+This release adds receipt and log saving functionality to the client. There has been a new database infrastructure laid out in PR [#1556](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1556) for data, caching and indexes.
+
+A new `ReceiptsManager` handles receipt and log saving and lookup, along with the `txHash -> [blockHash, txIndex]` index that is limited by `--txLookupLimit` (default = 2350000 blocks = about one year, 0 = entire chain).
+
+Enable with `--saveReceipts` and use with --rpc to enable `eth_getLogs`, `eth_getTransactionReceipt` and `eth_getTransactionByHash` endpoints.
+
+### Sepolia PoW Testnet Support
+
+The client now support the new [Sepolia](https://sepolia.ethdevops.io/) testnet, which is a PoW network intended to replace the coming-to-ages `ropsten` network, see PR [#1581](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1581).
```
Comment (ryanio) [inline packages/client/CHANGELOG.md:21]:
```suggestion
The client now supports the new [Sepolia](https://sepolia.ethdevops.io/) testnet, which is a PoW network intended to replace the `ropsten` network, see PR [#1581](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1581).
```

Comment (ryanio) [reply packages/client/CHANGELOG.md:21]: didn't really understand what you meant by "coming-to-ages" 😄

File: packages/vm/CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,34 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 (modification: no type change headlines) and this project adheres to
 [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## 5.7.0 - 2022-01-20
+
+### Dynamic Gas Costs
+
+Jochem from our team did a great refactoring how the VM handles gas costs in PR [#1364](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1364) by splitting up the opcode gas cost calculation (new file: `evm/opcodes/gas.ts`) from their actual behavior (stack edits, getting block hashes, etc.).
+
+This initial work was adopted a bit in PR [#1553](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1553) to remain backwards-compatibility and now allows to output the dynamic gas cost value in the VM `step` event (see `README`) now taking things like memory usage, address access or storage changes into account and therefore much better reflecting the real gas usage than only showing the (much lower) static part.
```
Comment (ryanio) [inline packages/vm/CHANGELOG.md:15]:
```suggestion
This initial work was adopted a bit in PR [#1553](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1553) to remain backwards-compatible and now allows to output the dynamic gas cost value in the VM `step` event (see `README`) now taking things like memory usage, address access or storage changes into account and therefore much better reflecting the real gas usage than only showing the (much lower) static part.
```

File: packages/vm/CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,34 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+
+So along to the static `opcode.fee` output there is now a new event object property `opcode.dynamicFee`.
```
Comment (ryanio) [inline packages/vm/CHANGELOG.md:17]:
```suggestion
So along with the static `opcode.fee` output there is now a new event object property `opcode.dynamicFee`.
```

File: packages/vm/CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,34 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+
+### StateManager Refactoring
+
+The VM `StateManager` has been substantially refactored in PR [#1548](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1548) and most of the generic functionality has been extracted to a super class `BaseStateManager`. This should make it substantially easier to do custom `StateManager` implementations with an alternativ access to the state by inheriting from `BaseStateManager` and only adopt the methods which directly access the underlying data structure. Have a look at the existing `DefaultStateManager` implementation for some guidance.
```
Comment (ryanio) [inline packages/vm/CHANGELOG.md:21]:
```suggestion
The VM `StateManager` has been substantially refactored in PR [#1548](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1548) and most of the generic functionality has been extracted to a super class `BaseStateManager`. This should make it substantially easier to do custom `StateManager` implementations with an alternative access to the state by inheriting from `BaseStateManager` and only adopting the methods which directly access the underlying data structure. Have a look at the existing `DefaultStateManager` implementation for some guidance.
```

File: packages/client/CHANGELOG.md (hunk 1 above)
Comment (holgerd77) [reply packages/client/CHANGELOG.md:21]: That's a direct translation of the german "in die Jahre kommen" which basically means "getting old". 😋 Might not exist in english. 😅

**Comment by holgerd77** [comment]
Updated this branch, this is now ready for a final review. 🙂 

I am planning to do the releases on this tomorrow (Tuesday, Feb 1st 2022), please do not merge any new PRs until then.

Also: please do a somewhat "loose" review overlooking a bit on minor typos and the like, so that releases are not postponed by that. If you do have things which would need some correction it would be optimal if you could just update yourself and then give an approval.

**Review by ryanio — APPROVED** [review]
lgtm!

**Comment by holgerd77** [comment]
Just released the following versions on npm:

- `ethereumjs-util` [v7.1.4](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/ethereumjs-util%407.1.4)
- `@ethereumjs/common` [v2.6.1](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Fcommon%402.6.1)
- `@ethereumjs/devp2p` [v4.2.1](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Fdevp2p%404.2.1)
- `merkle-patricia-tree` [v4.2.3](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/merkle-patricia-tree%404.2.3)
- `@ethereumjs/tx`[v3.5.0](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Ftx%403.5.0)
- `@ethereumjs/vm` [v5.7.0](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Fvm%405.7.0)
- `@ethereumjs/client` [v0.3.0](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Fclient%400.3.0)

**Pull request merged**